spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/FirestoreTemplate.java

```diff
@@ -81,6 +92,39 @@ public FirestoreTemplate(FirestoreStub firestore, String parent) {
 		});
 	}
 
+	@Override
+	public <T> Flux<T> saveAll(Publisher<T> instances) {
+		Flux<T> input = Flux.from(instances);
+		return ObservableReactiveUtil.streamingBidirectionalCall(this::openWriteStreamObserver, input,
+				(T entity, Flux<WriteResponse> responses) -> {
+					Mono<WriteResponse> firstResponse = responses.next();
+					return firstResponse.map((WriteResponse streamIds) -> {
+
+						FirestorePersistentEntity<?> persistentEntity = this.mappingContext
+								.getPersistentEntity(entity.getClass());
+						FirestorePersistentProperty idProperty = persistentEntity.getIdPropertyOrFail();
+						Object idVal = persistentEntity.getPropertyAccessor(entity).getProperty(idProperty);
+
+						Map<String, Value> valuesMap = PublicClassMapper.convertToFirestoreTypes(entity);
+
+						return WriteRequest.newBuilder()
+								.setStreamId(streamIds.getStreamId())
+								.setStreamToken(streamIds.getStreamToken())
+								.addWrites(Write.newBuilder()
+										.setUpdate(Document.newBuilder()
+												.putAllFields(valuesMap)
+												.setName(this.parent + "/"
+														+ persistentEntity.collectionName() + "/"
+														+ idVal.toString())
+												.build())
+										.build())
+								.build();
+
+					});
+				})
+				.filter(response -> response.getWriteResultsCount() > 0).thenMany(input);
```

Comment:
What is this doing? Is it basically ignoring the flux of responses, and just emitting the input flux?
What's the point of doing filter then?
If you we don't want to transform `WriteResponse`s to something meaningful just yet, then maybe just return a `Flux<Integer>`, which would just be the flux of `response.getWriteResultsCount()` from all responses.

Reply:
The return type is dictated by the reactive crud repo interface we ultimately support.

The chaining is absolutely necessary as it causes the earlier steps of the chain to actually execute.

Reply:
The filter is how we distinguish write confirmations from other misc replies.

Only write confirmations fill that condition.

Reply:
@meltsufin

Reply:
Maybe I don't understand `thenMany()` correctly, but it seems to me regardless of what happens with the response flux, as soon as it completes, the whole input flux is emitted. Am I missing something?

Reply:
Actually, that's right; I missed this. The response we need to filter out is the first one, so filtering it out won't affect when the Flux completes.
As far as whether to stream the output -- if Firestore sends back a stream of individual update statuses, then I agree with Mike that we should return the dynamic flux to the end user, who is then free to apply chaining operations like `thenMany`. But I am not sure how to interpret Firestore docs -- will the response actually be a stream of per-write responses, or will there be one WriteResponse containing an array of write statuses?

Reply:
Never mind the last part; I see that it's done on each signal from request flux.